### PR TITLE
cmd, core, eth, les, node: chain freezer on top of db rework

### PR DESCRIPTION
**Pull request by karalabe**
This PR implements a long desired feature of moving ancient chain segments out of the active database into immutable append-only flat files.

There are a few goals of this new mechanism:

By storing less data in the active key/value database, compactions should in theory be a lot faster since there's less to shuffle around. In practice we already prefixed chain items with their block numbers, so compactions might not benefit too much in themselves.
By storing less data in the active key/value database, the same amount of cache and file descriptor allowance should provide better performance, since more state trie nodes can be kept in active memory.
By moving the ancient chain items into their own flat files, we can split the database into an active set requiring a fast SSD disk and an immutable set for which a slow HDD disk suffices. This should permit more people to run Ethereum nodes without huge SSDs (--datadir.ancient=/path/on/hdd).
By keeping ancient chain items in flat files, we can piggyback on the operating system's built in memory manager and file cacher, allowing Geth to potentially max out available memory without actually having to hog it. This should reduce the strain on Go's GC, reduce the amount of memory used by Geth, while at the same time increase overall system stability.

### CODE REVIEW

File: core/rawdb/accessors_chain.go [hunk 1]
```diff
@@ -51,6 +54,24 @@ func DeleteCanonicalHash(db ethdb.Deleter, number uint64) {
 	}
 }
 
+// readAllHashes retrieves all the hashes assigned to blocks at a certain heights,
```
Comment (ligi) [inline core/rawdb/accessors_chain.go]:
```suggestion
// readAllHashes retrieves all the hashes assigned to blocks at certain heights,
```

File: core/rawdb/database.go [hunk 2]
```diff
@@ -22,10 +22,44 @@ import (
 	"github.com/ethereum/go-ethereum/ethdb/memorydb"
 )
 
+// freezerdb is a databse wrapper that enabled freezer data retrievals.
```
Comment (ligi) [inline core/rawdb/database.go]:
```suggestion
// freezerdb is a databse wrapper that enables freezer data retrievals.
```

File: core/rawdb/freezer.go [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+//
+// This functionality is deliberately broken off from block importing to avoid
+// incurring additional data shuffling delays on block propagation.
+func (f *freezer) freeze(db ethdb.KeyValueStore) {
+	nfdb := &nofreezedb{KeyValueStore: db}
+
+	for {
+		// Retrieve the freezing threshold. In theory we're interested only in full
+		// blocks post-sync, but that would keep the live database enormous during
+		// dast sync. By picking the fast block, we still get to deep freeze all the
```
Comment (Matthalp) [inline core/rawdb/freezer.go]: s/dast/fast

File: core/rawdb/freezer.go [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+				break
+			}
+			td := ReadTdRLP(nfdb, hash, f.frozen)
+			if len(td) == 0 {
+				log.Error("Total difficulty missing, can't freeze", "number", f.frozen, "hash", hash)
+				break
+			}
+			// Inject all the components into the relevant data tables
+			if err := f.tables["hashes"].Append(f.frozen, hash[:]); err != nil {
+				log.Error("Failed to deep freeze hash", "number", f.frozen, "hash", hash, "err", err)
```
Comment (Matthalp) [inline core/rawdb/freezer.go]: It seems like that if one of the subsequent `freezerTable.Append`s fails and then the import is retried (without reopening the freezer) then this will result in a panic since the entry will already have been written? Is that the failure model you want?

File: core/rawdb/freezer.go [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+			log.Crit("Failed to delete frozen items", "err", err)
+		}
+		// Log something friendly for the user
+		context := []interface{}{
+			"blocks", f.frozen - first, "elapsed", common.PrettyDuration(time.Since(start)), "number", f.frozen - 1,
+		}
+		if n := len(ancients); n > 0 {
+			context = append(context, []interface{}{"hash", ancients[n-1]}...)
+		}
+		log.Info("Deep froze chain segment", context...)
```
Comment (Matthalp) [inline core/rawdb/freezer.go]: This is another example of something I would have extracted out to not distract from the core functionality in this method.

File: core/rawdb/freezer_table.go [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/golang/snappy"
```
[the remark this answers is not in the document]
Comment (holiman) [reply core/rawdb/freezer_table.go]: From an earlier test

16 blocks. 

- as one binary file, uncompressed `332K` ,compressed `254K`. 
- As 16 indivudually compressed files, `257K`

Afaik, we haven't tested that on headers

[the remark this answers is not in the document]
Comment (karalabe) [reply core/rawdb/freezer_table.go]: pic

[the remark this answers is not in the document]
Comment (karalabe) [reply core/rawdb/freezer_table.go]: Though we cut off a lot of redundant data from the receipts, so I expect snappy to be less effective since the charts were plotted.

File: core/rawdb/freezer_table.go [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+		return nil, errOutOfBounds
+	}
+	// Item reachable, retrieve the data content boundaries
+	offset := make([]byte, 8)
+	if _, err := t.offsets.ReadAt(offset, int64(item*8)); err != nil {
+		return nil, err
+	}
+	start := binary.LittleEndian.Uint64(offset)
+
+	if _, err := t.offsets.ReadAt(offset, int64((item+1)*8)); err != nil {
```
Comment (holiman) [inline core/rawdb/freezer_table.go]: I think this means that you can't read back the last item, since trying to read the `item +1` index will fail

File: core/rawdb/freezer_table.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Matthalp) [reply core/rawdb/freezer_table.go]: Storage-wise it seems worthwhile. Is the decompression overhead for reads something to consider?

File: core/rawdb/freezer_table.go (hunk 7 above)
Comment (holiman) [reply core/rawdb/freezer_table.go]: nevermind, my bad

File: core/rawdb/freezer_table.go [hunk 8]
```diff
@@ -0,0 +1,493 @@
...
+	// errOutOfBounds is returned if the item requested is not contained within the
+	// freezer table.
+	errOutOfBounds = errors.New("out of bounds")
+)
+
+// indexEntry contains the number/id of the file that the data resides in, aswell as the
+// offset within the file to the end of the data
+// In serialized form, the filenum is stored as uint16.
+type indexEntry struct {
+	filenum uint32 // stored as uint16 ( 2 bytes)
```
Comment (Matthalp) [inline core/rawdb/freezer_table.go]: Is the index's storage footprint such a concern that you have to use `uint[16|32]` instead of `uint64`? Using `uint64` seems simpler, consistent with the `BlockChain` API, and is the common word size in modern processors. It just feels like an additional piece of context to remember without having any substantial impact on a running system.

File: core/rawdb/freezer_table.go [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+	readMeter  metrics.Meter // Meter for measuring the effective amount of data read
+	writeMeter metrics.Meter // Meter for measuring the effective amount of data written
+
+	logger log.Logger   // Logger with database path and table name ambedded
+	lock   sync.RWMutex // Mutex protecting the data file descriptors
+}
+
+// newTable opens a freezer table with default settings - 2G files and snappy compression
+func newTable(path string, name string, readMeter metrics.Meter, writeMeter metrics.Meter) (*freezerTable, error) {
+	return newCustomTable(path, name, readMeter, writeMeter, 2*1000*1000*1000, false)
```
Comment (Matthalp) [inline core/rawdb/freezer_table.go]: Nit: Consider making this a constant. Also should this be `2 * 1024 * 1024 * 1024`?

File: core/rawdb/freezer_table.go [hunk 10]
```diff
@@ -0,0 +1,493 @@
...
+// preopen opens all files that the freezer will need. This method should be called from an init-context,
+// since it assumes that it doesn't have to bother with locking
+// The rationale for doing preopen is to not have to do it from within Retrieve, thus not needing to ever
+// obtain a write-lock within Retrieve.
+func (t *freezerTable) preopen() (err error) {
+	// The repair might have already opened (some) files
+	t.releaseFilesAfter(0, false)
+	// Open all except head in RDONLY
+	for i := uint32(0); i < t.headId; i++ {
+		if _, err = t.openFile(i, os.O_RDONLY); err != nil {
```
Comment (Matthalp) [inline core/rawdb/freezer_table.go]: Are there any issues with file handler management here? The scenario I'm imagining is when multiple tables each have a bunch of files open (and LevelDB is also active).

File: core/rawdb/freezer_table.go [hunk 11]
```diff
@@ -0,0 +1,493 @@
...
+	t.head, err = t.openFile(t.headId, os.O_RDWR|os.O_CREATE|os.O_APPEND)
+	return err
+}
+
+// truncate discards any recent data above the provided threashold number.
+func (t *freezerTable) truncate(items uint64) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	// If out item count is corrent, don't do anything
+	if atomic.LoadUint64(&t.items) <= items {
```
Comment (Matthalp) [inline core/rawdb/freezer_table.go]: Is the `atomic.LoadUint64` needed here if the write lock has already been acquired?

File: core/rawdb/freezer_table.go [hunk 12]
```diff
@@ -0,0 +1,493 @@
...
+// Assumes that the caller holds the write lock
+func (t *freezerTable) releaseFile(num uint32) {
+	if f, exist := t.files[num]; exist {
+		delete(t.files, num)
+		f.Close()
+	}
+}
+
+// releaseFilesAfter closes all open files with a higher number, and optionally also deletes the files
+func (t *freezerTable) releaseFilesAfter(num uint32, remove bool) {
```
Comment (Matthalp) [inline core/rawdb/freezer_table.go]: Nit: I find the `remove` option to be misleading for a method about releasing. I would consider adding a second method that makes this more explicit. They can still use a common helper function.

File: core/rawdb/freezer_table.go [hunk 13]
```diff
@@ -0,0 +1,493 @@
...
+// fsync before irreversibly deleting data from the database.
+func (t *freezerTable) Append(item uint64, blob []byte) error {
+	// Read lock prevents competition with truncate
+	t.lock.RLock()
+	// Ensure the table is still accessible
+	if t.index == nil || t.head == nil {
+		return errClosed
+	}
+	// Ensure only the next item can be written, nothing else
+	if atomic.LoadUint64(&t.items) != item {
```
Comment (Matthalp) [inline core/rawdb/freezer_table.go]: Is the `atomic.LoadUint64` necessary if you've already acquired a read lock?

File: core/rawdb/freezer_table.go [hunk 14]
```diff
@@ -0,0 +1,493 @@
...
+
+// indexEntry contains the number/id of the file that the data resides in, aswell as the
+// offset within the file to the end of the data
+// In serialized form, the filenum is stored as uint16.
+type indexEntry struct {
+	filenum uint32 // stored as uint16 ( 2 bytes)
+	offset  uint32 // stored as uint32 ( 4 bytes)
+}
+
+const indexEntrySize = 6
```
[the remark this answers is not in the document]
Comment (karalabe) [reply core/rawdb/freezer_table.go]: Whilst I generally agree with that, this number depends on the above type declaration. Moving it elsewhere would give the false sense that it can be modified freely.

File: core/rawdb/freezer_table.go (hunk 8 above)
Comment (karalabe) [reply core/rawdb/freezer_table.go]: In-memory wise it doesn't matter tbh, it could even be a plain int. On-disk storage wise we know it won't ever exceed `uint16`, so no point in wasting more disk space on zeroes.

File: core/rawdb/freezer_table.go (hunk 9 above)
Comment (karalabe) [reply core/rawdb/freezer_table.go]: Yeah, I did wonder about this. I think I'd go with your suggestion.

Comment (karalabe) [reply core/rawdb/freezer_table.go]: We just need to take extra care not to ever overflow it, since many file systems + 32bit architectures are hard limited to that number.

Comment (karalabe) [reply core/rawdb/freezer_table.go]: Ah, scratch 32 bit arch, that should handle 4GB (or 3 in case of windows, lol).

File: core/rawdb/freezer_table.go (hunk 10 above)
Comment (karalabe) [reply core/rawdb/freezer_table.go]: The current historical chain is about 80GB in size, so that's +- 40 fds. These will probably be opened either way if someone syncs from us. I think 40 is safe, but if we grow we probably want to eventually add fd management. That seems an overcomplication at this point however.

File: core/rawdb/freezer_table.go (hunk 11 above)
Comment (karalabe) [reply core/rawdb/freezer_table.go]: The locking mechanism was designed to allow reading and appending to be concurrently runable (since append is just pushes some data that readers are unaware of, and at the very end bumps the item count, making it available).

If append does not write lock, just uses atomic, then this write lock here doesn't protect the `items` counter. Append will only obtain a write lock if it needs to switch to a new file. That said, Append and Truncate may not really be safe to use concurrently, as they might race on what the final value of items should be.

File: core/rawdb/freezer_table.go (hunk 13 above)
Comment (karalabe) [reply core/rawdb/freezer_table.go]: Yes, Append can run concurrently (since it doesn't change internals, just adds something to the end of the file and them atomically makrs it available).

File: core/rawdb/freezer_table.go (hunk 8 above)
Comment (Matthalp) [reply core/rawdb/freezer_table.go]: I hear you. Just figured I would point it out.

File: core/rawdb/freezer_table.go (hunk 10 above)
Comment (Matthalp) [reply core/rawdb/freezer_table.go]: Agree. I was just asking to serve as a sanity check.

File: core/rawdb/freezer_table.go (hunk 11 above)
Comment (Matthalp) [reply core/rawdb/freezer_table.go]: If it were me I would be more conservative and lock more frequently -- especially for writes. I feel like any slow down would be negligible and it ensures the code is more safe. Just my two wei.

File: core/rawdb/freezer_table.go (hunk 9 above)
Comment (holiman) [reply core/rawdb/freezer_table.go]: Meh. ISO standards FTW

**Comment by karalabe** [comment]
Nitpick to do: make data file numbers zero padded to 3 digits. 

TL;DR

```
-rw-r--r-- 1 root root 1999998703 Mar 27 16:37 receipts.0.cdat
-rw-r--r-- 1 root root 1999997444 Mar 27 16:54 receipts.1.cdat
-rw-r--r-- 1 root root 1999999967 Mar 27 19:48 receipts.10.cdat
-rw-r--r-- 1 root root 1999999260 Mar 27 20:08 receipts.11.cdat
-rw-r--r-- 1 root root 1999994597 Mar 27 20:30 receipts.12.cdat
-rw-r--r-- 1 root root 1999993520 Mar 27 20:51 receipts.13.cdat
-rw-r--r-- 1 root root 1999992762 Mar 27 21:12 receipts.14.cdat
-rw-r--r-- 1 root root 1999990541 Mar 27 21:33 receipts.15.cdat
-rw-r--r-- 1 root root 1999988154 Mar 27 21:53 receipts.16.cdat
-rw-r--r-- 1 root root 1999995262 Mar 27 22:14 receipts.17.cdat
-rw-r--r-- 1 root root 1273323457 Mar 27 22:27 receipts.18.cdat
-rw-r--r-- 1 root root 1999989717 Mar 27 17:07 receipts.2.cdat
-rw-r--r-- 1 root root 1999995186 Mar 27 17:21 receipts.3.cdat
-rw-r--r-- 1 root root 1999999174 Mar 27 17:40 receipts.4.cdat
-rw-r--r-- 1 root root 1999990962 Mar 27 18:03 receipts.5.cdat
-rw-r--r-- 1 root root 1999995600 Mar 27 18:26 receipts.6.cdat
-rw-r--r-- 1 root root 1999997331 Mar 27 18:47 receipts.7.cdat
-rw-r--r-- 1 root root 1999993270 Mar 27 19:08 receipts.8.cdat
-rw-r--r-- 1 root root 1999998447 Mar 27 19:28 receipts.9.cdat
```

**Comment by holiman** [comment]
Here are some charts from the test with @rjl493456442 's changes that was run in fast-sync. 

This is the master. It has a very intense download of blocks/headers/state for about 4 hours, then a long drawn-out state download which takes roughly another 8 hours. 
Screenshot_2019-03-28 EXP (52 56 200 108) VS Master (52 56 85 219) Datadog

For the experimental, it looks a bit different. It has an 'intense' download lasting six hours, followed by one hours of less intense download, and finishes the fast sync in a total of 7 hours.
Screenshot_2019-03-28 EXP (52 56 200 108) VS Master (52 56 85 219) Datadog(1)

So 7 hours instead of 12. I'm not quite sure _why_. 


The malloc/gc overhead was more stable on experimental:

Screenshot_2019-03-28 EXP (52 56 200 108) VS Master (52 56 85 219) Datadog(2)

**Comment by holiman** [comment]
More stats on the data
```
root@mon07:/datadrive/geth/geth/chaindata# du -h .
88G    ./ancient
153G    .
root@mon06:/datadrive/geth/geth/chaindata# du -h .
154G    .
```
So the `master` levedb stores `154G` of data, whereas the `experimental` has `65G` in levedb and `88G` in the `ancient`

**Comment by Matthalp** [comment]
@holiman The results make sense to me. There is less pressure on LevelDB since it's now responsible for handling half the amount of data as before, which reduces its operation overhead (e.g. garbage collection). The LevelDB storage hierarchy is also organized by powers of 10, so going from ~150 GB to half that means that one less level is needed.

**Comment by Matthalp** [comment]
Nonetheless it's probably worth running it a few times. I'm not sure the exact benchmark setup but if a different peer is being used for each sync then that could introduce some variation as well.

**Comment by holiman** [comment]
Yeah - we did just that. We swapped the machines, wiped the node-ids and ran again. It's still running, but this time it looks like the `master` will finish ahead of this PR. So that was probably a one-off thing - the big difference for statesync is the quality of your peers.

**Comment by holiman** [comment]
So on the second run, they were roughly equal -- this PR making a fast sync in ~10h, the `master` half an hour faster. Also, on the second run, the difference in mallocs and gc was not as pronounced:

Screenshot_2019-03-29 EXP (52 56 200 108) VS Master (52 56 85 219) Datadog

**OBS** Since we switched machines, the colours are swapped in that ^ chart, this PR is the blue line.

File: cmd/geth/chaincmd.go [hunk 15]
```diff
@@ -367,9 +367,12 @@ func exportPreimages(ctx *cli.Context) error {
 
 func copyDb(ctx *cli.Context) error {
 	// Ensure we have a source chain directory to copy
-	if len(ctx.Args()) != 1 {
+	if len(ctx.Args()) < 1 {
 		utils.Fatalf("Source chaindata directory path argument missing")
 	}
+	if len(ctx.Args()) < 2 {
```
[the remark this answers is not in the document]
Comment (karalabe) [reply cmd/geth/chaincmd.go]: By default yes, but if we want to copydb from a database where the ancient is outside?

File: core/blockchain.go [hunk 16]
```diff
@@ -782,96 +833,259 @@ func (bc *BlockChain) Rollback(chain []common.Hash) {
...
+			// Short circuit insertion if it is required(used in testing only)
+			if bc.terminateInsert != nil && bc.terminateInsert(block.Hash(), block.NumberU64()) {
+				return i, errors.New("insertion is terminated for testing purpose")
+			}
+			// Short circuit if the owner header is unknown
+			if !bc.HasHeader(block.Hash(), block.NumberU64()) {
+				return i, fmt.Errorf("containing header #%d [%x…] unknown", block.Number(), block.Hash().Bytes()[:4])
+			}
+			// Compute all the non-consensus fields of the receipts
+			if err := receiptChain[i].DeriveFields(bc.chainConfig, block.Hash(), block.NumberU64(), block.Transactions()); err != nil {
```
Comment (rjl493456442) [inline core/blockchain.go]: We don't need to derive the receipt fields here since they will be discarded during the RLP encoding anyway.

File: core/blockchain.go [hunk 17]
```diff
@@ -782,96 +833,259 @@ func (bc *BlockChain) Rollback(chain []common.Hash) {
...
+			// Short circuit if the owner header is unknown
+			if !bc.HasHeader(block.Hash(), block.NumberU64()) {
+				return i, fmt.Errorf("containing header #%d [%x…] unknown", block.Number(), block.Hash().Bytes()[:4])
+			}
+			if bc.HasBlock(block.Hash(), block.NumberU64()) {
+				stats.ignored++
+				continue
+			}
+			// Compute all the non-consensus fields of the receipts
+			if err := receiptChain[i].DeriveFields(bc.chainConfig, block.Hash(), block.NumberU64(), block.Transactions()); err != nil {
```
Comment (rjl493456442) [inline core/blockchain.go]: We don't need to derive the receipt fields here since they will be discarded during the RLP encoding anyway.

File: core/rawdb/freezer.go [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+		table, err := newTable(datadir, name, readMeter, writeMeter)
+		if err != nil {
+			for _, table := range freezer.tables {
+				table.Close()
+			}
+			return nil, err
+		}
+		freezer.tables[name] = table
+	}
+	// Truncate all data tables to the same length
```
[the remark this answers is not in the document]
Comment (hadv) [reply core/rawdb/freezer.go]: loving your usage of `my two wei` 👍

File: core/rawdb/freezer.go [hunk 19]
```diff
@@ -0,0 +1,365 @@
...
+	atomic.AddUint64(&f.frozen, 1) // Only modify atomically
+	return nil
+}
+
+// Truncate discards any recent data above the provided threshold number.
+func (f *freezer) TruncateAncients(items uint64) error {
+	if atomic.LoadUint64(&f.frozen) <= items {
+		return nil
+	}
+	for _, table := range f.tables {
```
Comment (rjl493456442) [inline core/rawdb/freezer.go]: We also need the `repair` mechanism here to ensure all tables are synced.

File: core/rawdb/freezer_table.go [hunk 20]
```diff
@@ -0,0 +1,548 @@
...
+
+		// Swap out the current head
+		t.head = newHead
+		atomic.StoreUint32(&t.headBytes, 0)
+		atomic.StoreUint32(&t.headId, nextId)
+		t.lock.Unlock()
+		t.lock.RLock()
+	}
+
+	defer t.lock.RUnlock()
```
Comment (rjl493456442) [inline core/rawdb/freezer_table.go]: The lock is used to protect `map[uint32]*os.File` as well as `head` and `index` file descriptor. Since `os.File.Write` is not thread safe, should we use `Lock` here instead of `RLock`?

File: eth/downloader/downloader.go [hunk 21]
```diff
@@ -464,12 +465,49 @@ func (d *Downloader) syncWithPeer(p *peerConnection, hash common.Hash, td *big.I
...
+		// checkpoint. Otherwise calculate the ancient limit through the advertised
+		// height by remote peer.
+		//
+		// The reason for picking checkpoint first is: there exists an attack vector
+		// for height that: a malicious peer can give us a fake(very high) height,
+		// so that the ancient limit is also very high. And then the peer start to
+		// feed us valid blocks until head. All of these blocks might be written into
+		// the ancient store, the safe region for freezer is not enough.
+		if d.checkpoint != 0 && d.checkpoint > MaxForkAncestry+1 {
+			d.ancientLimit = height - MaxForkAncestry - 1
```
Comment (rjl493456442) [inline eth/downloader/downloader.go]: `d.ancientLimit = height - MaxForkAncestry - 1` -> `d.ancientLimit = d.checkpoint - MaxForkAncestry - 1`

File: core/rawdb/freezer_table.go (hunk 20 above)
Comment (holiman) [reply core/rawdb/freezer_table.go]: I think maybe we should add a freezer.md where we explain things like the concurrency assumptions and the data layout. To answer the question, I think No; we do not cater for multiple callers of `Append`, tha aim is to prevent concurrent calling of `Append/Truncate`.

**Comment by holiman** [comment]
Data taken from `mon07`, comparing the effectiveness of snappy versus encoding plan. 
receipts
headers
hashes
diffs
bodies

Sorry for the crappyness of the labeling, but aside from that, in order of size:

- bodies: `76 e9` versus `55 e9` (snappy wins)
- receipts `61 e9` versus `22 e9` (snappy wins)
- hashes `260 e6` versus `240 e6` (plain would be smaller)
- difficulty `91 e6`versus `75 e6` (plain would be smaller). 
- headers `4.1 e9` versus `3.2 e9` (snappy wins)

Also, the index for each type is takes rougly `44MB` on disk each.

File: core/rawdb/freezer.go [hunk 22]
```diff
@@ -0,0 +1,365 @@
...
+			continue
+
+		case *number < freezerBlockGraduation:
+			log.Debug("Current full block not old enough", "number", *number, "hash", hash, "delay", freezerBlockGraduation)
+			time.Sleep(freezerRecheckInterval)
+			continue
+
+		case *number-freezerBlockGraduation <= f.frozen:
+			log.Debug("Ancient blocks frozen already", "number", *number, "hash", hash, "frozen", f.frozen)
+			time.Sleep(freezerRecheckInterval)
```
Comment (Matthalp) [inline core/rawdb/freezer.go]: Should this be an error?

**Comment by holiman** [comment]
TODO, as far as I know

[[ ]] - Make sure that if you use an `ancient` database which already has data, but an otherwise cliean leveldb, that things at least 'work'.

**Comment by adamschmideg** [comment]
Waiting for https://github.com/karalabe/go-ethereum/pull/15 to be merged

**Comment by Matthalp** [comment]
@karalabe Did you consider using separate LevelDB instances to store blockchain constructs (e.g. what each `Table` is storing) rather than using a freezer? I think it will accomplish most of the goals of this implementation without all of the implementation complexity. It seems like it would be pretty straightforward to do inside of #19200.

**Comment by karalabe** [comment]
@Matthalp This PR did get a bit more complex than I anticipated, though that's because we decided to split the data files into 2GB chunks. Without that the code was really really simple.

That said, the operations we are doing is relatively straightforward. The complexity comes mostly from optimizations. With our current code we have a very nice control over how much resources we waste (0 memory, 10-20 file descriptors vs. caches + files for leveldb), and we control how much computational overhead we have (0 vs. compaction for leveldb).

Yes, leveldb might be a tad nicer, but by rolling our own files we can take advantage of all the properties of the data we are storing and optimize for our specific use case.

File: core/blockchain.go [hunk 23]
```diff
@@ -961,16 +1006,38 @@ func (bc *BlockChain) InsertReceiptChain(blockChain types.Blocks, receiptChain [
...
+					break
+				}
+				h := rawdb.ReadCanonicalHash(bc.db, frozen)
+				b := rawdb.ReadBlock(bc.db, h, frozen)
+				size += rawdb.WriteAncientBlock(bc.db, b, rawdb.ReadReceipts(bc.db, h, frozen, bc.chainConfig), rawdb.ReadTd(bc.db, h, frozen))
+				count += 1
+
+				// Always keep genesis block in active database.
+				if b.NumberU64() != 0 {
+					deleted = append(deleted, b)
```
Comment (holiman) [inline core/blockchain.go]: This can cause a _huge_ spike in memory, our monitoring node spiked at nearly `40G` memory shoving four million blocks from leveldb to ancients

**Comment by holiman** [comment]
No, there is no such switch. It should be fairly trivial to disable the live conversion from leveldb to freezer, but there's also the initial download-to-freezer which is maybe a bit less trivial. 
Why would you want to disable it though?

**Comment by 317787106** [comment]
@holiman, I want to konw the space-saving rate about the conversion from leveldb to freezer. If i disable the freezer firstly and then enable that feature, we can calculate the rate of space. Is there any experimental data?
